Span Symmetric ICryptoTransform

vcsjones opened the pull request:
This changes the underlying `UniversalCryptoTransform`, derived types, and symmetric cryptographic primitives to operate on span.

This gains some performance benefits already, either by avoiding temporary arrays (removing and adding padding) and using `CryptoPool` where possible.

~This also would pave the way for `{ReadOnly}Span<byte>` to `ICryptoTransform` and the built-in primitives, provided that a sensible default behavior is possible with DIM.~

These changes are also motivated by #2406, which will operate on span.

<details>
<summary>Performance Numbers</summary>
<pre>
|                      Method |        Job |   Toolchain | DataSize |  Mode |            Mean |         Error |        StdDev |          Median | Ratio | RatioSD |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|---------------------------- |----------- |------------ |--------- |------ |----------------:|--------------:|--------------:|----------------:|------:|--------:|---------:|---------:|---------:|----------:|
|      Encrypt_TransformBlock | Job-QQUEGF | crypto-span |  1048576 | PKCS7 | 1,026,981.62 ns |  9,912.516 ns |  8,787.188 ns | 1,029,210.30 ns |  1.00 |    0.01 |        - |        - |        - |       1 B |
|      Encrypt_TransformBlock | Job-PLPBZT |      master |  1048576 | PKCS7 | 1,024,755.20 ns |  3,505.305 ns |  3,278.864 ns | 1,024,598.03 ns |  1.00 |    0.00 |        - |        - |        - |       1 B |
|                             |            |             |          |       |                 |               |               |                 |       |         |          |          |          |           |
|      Decrypt_TransformBlock | Job-QQUEGF | crypto-span |  1048576 | PKCS7 |   135,297.81 ns |    453.515 ns |    424.218 ns |   135,393.02 ns |  0.98 |    0.01 |        - |        - |        - |         - |
|      Decrypt_TransformBlock | Job-PLPBZT |      master |  1048576 | PKCS7 |   138,163.82 ns |    687.527 ns |    609.475 ns |   138,280.70 ns |  1.00 |    0.00 |        - |        - |        - |         - |
|                             |            |             |          |       |                 |               |               |                 |       |         |          |          |          |           |
| Encrypt_TransformFinalBlock | Job-QQUEGF | crypto-span |  1048576 | PKCS7 | 1,225,488.74 ns | 11,252.529 ns |  9,975.074 ns | 1,226,682.06 ns |  0.80 |    0.01 | 195.3125 | 195.3125 | 195.3125 | 1048668 B |
| Encrypt_TransformFinalBlock | Job-PLPBZT |      master |  1048576 | PKCS7 | 1,531,450.18 ns | 12,607.728 ns | 10,528.024 ns | 1,534,052.30 ns |  1.00 |    0.00 | 250.0000 | 250.0000 | 250.0000 | 2097299 B |
|                             |            |             |          |       |                 |               |               |                 |       |         |          |          |          |           |
| Decrypt_TransformFinalBlock | Job-QQUEGF | crypto-span |  1048576 | PKCS7 |   806,700.19 ns | 11,034.824 ns | 10,321.981 ns |   808,663.70 ns |  0.73 |    0.03 | 397.4609 | 397.4609 | 397.4609 | 3145944 B |
| Decrypt_TransformFinalBlock | Job-PLPBZT |      master |  1048576 | PKCS7 | 1,111,906.51 ns | 21,502.634 ns | 27,193.948 ns | 1,121,153.26 ns |  1.00 |    0.00 | 269.5313 | 269.5313 | 269.5313 | 3145905 B |

</pre>
</details>

vcsjones committed:
Spanify PadBlock and use pool for padding block.

vcsjones committed:
Span TransformFinal.

Still creates arrays where the API shape currently demands it.

vcsjones committed:
Avoid temp buffer in TransformFinal.

If the input and output buffer have zero overlap, or if the
input buffer is effectively the same as the output buffer, we can
write directly to the output buffer.

vcsjones committed:
Spanify DepadBlock

vcsjones committed:
Spanify TransformBlock.

vcsjones committed:
Fix padding size.

vcsjones committed:
Get Apple building and passing.

vcsjones committed:
Get rid of the buffer and write to the output buffer.

vcsjones committed:
Span cryptographic transforms for CNG/CAPI.

vcsjones committed:
Span UncheckedTransformBlock.

vcsjones committed:
Spanify UncheckedTransformFinalBlock

vcsjones committed:
Don't create a temp buffer if we don't have held over data.

vcsjones committed:
Split encrypt and decrypt implementations.

When encrypting, we know the final size of the ciphertext, so we
can avoid a buffer and copy.

vcsjones committed:
Avoid pool and copy if no padding removal is required.

vcsjones committed:
Fix missing zeroing

vcsjones committed:
Combine AsSpan and Slice

vcsjones committed:
Some cleanup

vcsjones commented on src/libraries/System.Security.Cryptography.Csp/src/Resources/Strings.resx:
```diff
@@ -1,16 +1,16 @@
-  <!-- 
-    Microsoft ResX Schema 
-    
+  <!--
```
Sorry for the noise here, can undo if its undesirable in this PR, I'm assuming some editor config setting is removing the excessive white space.

vcsjones commented:
Presumably the AllocateUnitializedArray calls need a `#if NETCOREAPP5_0` guard and just new-up a regular array otherwise. Will address.

vcsjones committed:
Use uninitialized arrays only for .NET 5.

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs:
```diff
@@ -82,16 +82,21 @@ protected override unsafe int UncheckedTransformFinalBlock(ReadOnlySpan<byte> in
-                // Pinned so that we can avoid the `fixed` to stop the decrypt buffer from getting
-                // copied around by the GC.
-                ciphertext = GC.AllocateUninitializedArray<byte>(inputBuffer.Length, pinned: true);
+#if NET5_0
```
Rather than defining this as an opt-in to new, define it as an opt-in to old (e.g. `#if NETSTANDARD`).  Otherwise when we bump 5 to 6 this will fall to the "legacy" path.

vcsjones replied:
Gotcha. Fixed, I think.

vcsjones committed:
Flip ifdefs.

Opt-in to old behavior instead of opt-in to new behavior.

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs:
```diff
@@ -24,7 +24,7 @@ public UniversalCryptoDecryptor(PaddingMode paddingMode, BasicSymmetricCipher ba
         }
 
-        protected sealed override int UncheckedTransformBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
+        protected override int UncheckedTransformBlock(ReadOnlySpan<byte> inputBuffer, Span<byte> outputBuffer)
```
Should/can this one be sealed, like its predecessor?

vcsjones replied:
The whole class is sealed, wouldn't that make it redundant when specified on the method?

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/UniversalCryptoDecryptor.cs:
```diff
@@ -52,72 +52,111 @@ protected sealed override int UncheckedTransformBlock(byte[] inputBuffer, int in
+#if NETSTANDARD || NETFRAMEWORK || NETCOREAPP3_0
+                ciphertext = new byte[inputBuffer.Length];
+#else
+                ciphertext = GC.AllocateUninitializedArray<byte>(inputBuffer.Length);
```
We're not returning these arrays or calling back into callbacks, perhaps CryptoPool renting would be better?

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/UniversalCryptoEncryptor.cs:
```diff
@@ -23,81 +23,130 @@ public UniversalCryptoEncryptor(PaddingMode paddingMode, BasicSymmetricCipher ba
         {
-            byte[] paddedBlock = PadBlock(inputBuffer, inputOffset, inputCount);
+            int padWritten = PadBlock(inputBuffer, outputBuffer);
+            int transformWritten = BasicSymmetricCipher.TransformFinal(outputBuffer.Slice(0, padWritten), outputBuffer);
```
There are now two different times that the output buffer gets written to.  This could break something like

```C#
byte[] buf = new byte[1024];
Span<byte> dest = buf;
Span<byte> input = buf.Slice(41);
Fill(input);
Encrypt(input, dest)
```

If the padding happened to do an overwrite of the input buffer.

This could either use MemoryMarshal.Overlaps to see if there's enough space for a block of padding without contamination then falling back to a rented buffer, or just use a rented buffer.

Always renting a buffer is certainly an easier code-flow.

vcsjones replied:
Yeah, there is only one caller of this right now which is `byte[] UncheckedTransformFinalBlock(...)` and it is always guaranteed to pass in the buffer of the right size. I had presumed that "Unchecked" in the name was sufficient to indicate "Hey, you need to make sure you are passing in the right size", but if you think this API needs to be able to stand on its own, I can add something there that make sure there is enough room for padding.

It would just seem redundant right now for the one caller to say "this is the exactly right size you need" and the callee to rent a buffer anyway.

bartonjs replied:
So long as input and output don't overlap, this is fine.  Just throw in something like

```C#
// The only caller of this method is the array-allocating overload, outputBuffer is
// always new memory, not a user-provided buffer.
Debug.Assert(!inputBuffer.Overlaps(outputBuffer));
```

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/AppleCCCryptor.cs:
```diff
@@ -39,46 +39,36 @@ protected override void Dispose(bool disposing)
             int errorCode;
 
-            fixed (byte* outputStart = &output[0])
+            fixed (byte* outputStart = output)
```
Is it possible for us to get here with `output` being empty? If so, `outputStart` is 0/NULL, which might violate requirements of the called methods.

vcsjones replied:
It's not expected. This is asserted, just not straight forwardly.

A few lines up, we return early if `input.Length` is 0.

```
if (input.Length == 0)
{
    return 0;
}
```

So we know that input length is > 0. `ProcessFinalBlock` is called in exactly one place, by `TransformFinal`. It asserts:

```
Debug.Assert(input.Length <= output.Length);
```

So output's length by the time it reaches the `fixed` is `1 <= output.Length`.

I think I'll add an assert here anyway to spell all that out.

vcsjones committed:
Pool decrypt buffer

vcsjones committed:
Assert output size.

This could be reasoned about from other input checks, but added
to be clear that this is the expectation here.

bartonjs commented on src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs:
```diff
@@ -935,67 +923,47 @@ internal static void EncryptKey(SafeKeyHandle safeKeyHandle, byte[] pbKey, int c
-            int decryptedDataLength = inputCount;
+            Debug.Assert((input.Length % 8) == 0);
+
+            byte[] dataTobeDecrypted = new byte[input.Length];
```
```suggestion
            byte[] dataToBeDecrypted = new byte[input.Length];
```

bartonjs commented on src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs:
```diff
@@ -935,67 +923,47 @@ internal static void EncryptKey(SafeKeyHandle safeKeyHandle, byte[] pbKey, int c
-                outputCount = encryptedDataLength;
-            }
 
             // If isFinal, padding was added so ignore it by using outputCount as size
```
Since outputCount isn't defined yet, the comment could use a tweak.

bartonjs reviewed (commented):
Since I had already asked for an !overlaps assert, added a few nitpicky things that I was otherwise going to just be quiet about.

Looks good. Thanks for doing this.

vcsjones committed:
Fix variable name casing

vcsjones committed:
Add assert to ensure input / output buffers do not overlap.

vcsjones committed:
Fix comment to reflect changes.

vcsjones commented:
> added a few nitpicky things that I was otherwise going to just be quiet about.

I'm always happy to fix things!
